File: connector/ce/connector-ce/src/main/kotlin/de/sovity/edc/ce/api/ui/pages/asset/AssetApiService.java
/*
 * Copyright sovity GmbH and/or licensed to sovity GmbH under one or
 * more contributor license agreements. You may not use this file except
 * in compliance with the "Elastic License 2.0".
 *
 * SPDX-License-Identifier: Elastic-2.0
 */
package de.sovity.edc.ce.api.ui.pages.asset;

import de.sovity.edc.ce.api.common.model.UiAsset;
import de.sovity.edc.ce.api.common.model.UiAssetCreateRequest;
import de.sovity.edc.ce.api.common.model.UiAssetEditRequest;
import de.sovity.edc.ce.api.ui.model.IdResponseDto;
import de.sovity.edc.ce.api.ui.pages.dashboard.services.SelfDescriptionService;
import de.sovity.edc.ce.api.utils.ServiceException;
import de.sovity.edc.ce.db.jooq.Tables;
import de.sovity.edc.ce.libs.mappers.AssetMapper;
import de.sovity.edc.runtime.simple_di.Service;
import lombok.RequiredArgsConstructor;
import lombok.val;
import org.eclipse.edc.connector.controlplane.asset.spi.domain.Asset;
import org.eclipse.edc.connector.controlplane.services.spi.asset.AssetService;
import org.eclipse.edc.spi.query.QuerySpec;
import org.jetbrains.annotations.NotNull;
import org.jooq.DSLContext;

import java.util.Comparator;
import java.util.List;
import java.util.Objects;

@RequiredArgsConstructor
@Service
public class AssetApiService {
    private final AssetService assetService;
    private final AssetMapper assetMapper;
    private final AssetIdValidator assetIdValidator;
    private final SelfDescriptionService selfDescriptionService;

    public List<UiAsset> getAssets() {
        var assets = getAllAssets();
        var connectorEndpoint = selfDescriptionService.getConnectorEndpoint();
        var participantId = selfDescriptionService.getParticipantId();
        return assets.stream().sorted(Comparator.comparing(Asset::getCreatedAt).reversed())
            .map(asset -> assetMapper.buildUiAsset(asset, connectorEndpoint, participantId))
            .toList();
    }

    @NotNull
    public IdResponseDto createAsset(UiAssetCreateRequest request) {
        assetIdValidator.assertValid(request.getId());
        var organizationName = selfDescriptionService.getCuratorName();
        val asset = assetMapper.buildAsset(request, organizationName);
        val createdAsset = assetService.create(asset).orElseThrow(ServiceException::new);
        return new IdResponseDto(createdAsset.getId());
    }

    public IdResponseDto editAsset(String assetId, UiAssetEditRequest request) {
        val foundAsset = assetService.findById(assetId);
        Objects.requireNonNull(foundAsset, "Asset with ID %s not found".formatted(assetId));
        val editedAsset = assetMapper.editAsset(foundAsset, request);
        val updatedAsset = assetService.update(editedAsset).orElseThrow(ServiceException::new);
        assetService.update(editedAsset);
        return new IdResponseDto(updatedAsset.getId());
    }

    @NotNull
    public IdResponseDto deleteAsset(String assetId) {
        var response = assetService.delete(assetId).orElseThrow(ServiceException::new);
        return new IdResponseDto(response.getId());
    }

    private List<Asset> getAllAssets() {
        return assetService.search(QuerySpec.max()).orElseThrow(ServiceException::new);
    }

    public boolean assetExists(DSLContext dsl, String assetId) {
        val a = Tables.EDC_ASSET;
        return dsl.selectCount()
            .from(a)
            .where(a.ASSET_ID.eq(assetId))
            .fetchSingleInto(Integer.class) > 0;
    }
}

File: connector/ce/connector-ce/src/main/kotlin/de/sovity/edc/ce/api/ui/pages/asset/AssetApiService.kt
/*
 * Copyright sovity GmbH and/or licensed to sovity GmbH under one or
 * more contributor license agreements. You may not use this file except
 * in compliance with the "Elastic License 2.0".
 *
 * SPDX-License-Identifier: Elastic-2.0
 */
package de.sovity.edc.ce.api.ui.pages.asset

import de.sovity.edc.ce.api.common.model.UiAssetCreateRequest
import de.sovity.edc.ce.api.common.model.UiAssetEditRequest
import de.sovity.edc.ce.api.ui.model.IdResponseDto
import de.sovity.edc.ce.api.ui.pages.asset_detail_page.services.AssetDetailPageBuilder
import de.sovity.edc.ce.api.ui.pages.asset_detail_page.services.AssetDetailPageQueryService
import de.sovity.edc.ce.api.ui.pages.dashboard.services.SelfDescriptionService
import de.sovity.edc.ce.api.utils.jooq.EdcJsonUtils
import de.sovity.edc.ce.api.utils.notFoundError
import de.sovity.edc.ce.db.jooq.Tables
import de.sovity.edc.ce.libs.mappers.AssetMapper
import de.sovity.edc.ce.libs.mappers.asset.AssetJsonLdBuilder
import de.sovity.edc.runtime.simple_di.Service
import org.jooq.DSLContext
import org.jooq.JSON

@Service
class AssetApiService(
    private val assetMapper: AssetMapper,
    private val assetIdValidator: AssetIdValidator,
    private val selfDescriptionService: SelfDescriptionService,
    private val assetJsonLdBuilder: AssetJsonLdBuilder,
    private val assetDetailPageBuilder: AssetDetailPageBuilder,
    private val assetDetailPageQueryService: AssetDetailPageQueryService,
    private val jsonUtils: EdcJsonUtils
) {

    fun createAsset(dsl: DSLContext, request: UiAssetCreateRequest): IdResponseDto {
        assetIdValidator.assertValid(request.id)
        val organizationName = selfDescriptionService.curatorName
        val createAssetJson = assetJsonLdBuilder.buildCreateAssetJsonLds(request, organizationName)
        val at = Tables.EDC_ASSET
        dsl.newRecord(at).also {
            it.assetId = request.id
            it.createdAt = System.currentTimeMillis()
            it.dataAddress = JSON.json(createAssetJson.dataSource.toString())
            it.properties = JSON.json(createAssetJson.properties.toString())
            it.privateProperties = JSON.json(createAssetJson.privateProperties.toString())
            it.insert()
        }
        return IdResponseDto(request.id)
    }

    fun editAsset(dsl: DSLContext, assetId: String, request: UiAssetEditRequest): IdResponseDto {
        val a = Tables.EDC_ASSET
        val foundAsset =
            assetDetailPageQueryService.fetchAssetDetailPage(dsl, assetId)
                ?.let { assetDetailPageBuilder.buildAsset(it) }
                ?: notFoundError("Asset with ID $assetId not found")
        val editedAsset = assetMapper.editAsset(foundAsset, request)
        dsl.fetchOne(a, a.ASSET_ID.eq(assetId))?.also {
            it.dataAddress = jsonUtils.toPostgresqlJson(editedAsset.dataAddress.properties)
            it.properties = jsonUtils.toPostgresqlJson(editedAsset.properties)
            it.privateProperties = jsonUtils.toPostgresqlJson(editedAsset.privateProperties)
            it.update()
        }
        return IdResponseDto(assetId)
    }

    fun deleteAsset(dsl: DSLContext, assetId: String): IdResponseDto {
        val a = Tables.EDC_ASSET
        dsl.deleteFrom(a).where(a.ASSET_ID.eq(assetId)).execute()
        return IdResponseDto(assetId)
    }
}
